OMCompiler/Compiler/NFFrontEnd/NFFlatten.mo

```diff
@@ -2252,6 +2255,66 @@ algorithm
   execStat(getInstanceName());
 end resolveConnections;
 
+function generateTopLevelIOs
+  "generate top-level inputs and outputs for public unconnected local input and output connectors"
+  input UnorderedMap<ComponentRef, Variable> variables;
+  input UnorderedSet<ComponentRef> connectedIOs;
+  output list<Variable> tlio_vars;
+  output list<Equation> tlio_eql;
+protected
+  Attributes attributes;
+  Variable tlio_var;
+  ComponentRef cref;
+  String name;
+  InstNode tlio_node;
+algorithm
+  tlio_vars := {};
+  tlio_eql := {};
+  for variable in UnorderedMap.valueList(variables) loop
+    attributes := variable.attributes;
+    if variable.visibility == Visibility.PUBLIC and
+      attributes.connectorType <> ConnectorType.NON_CONNECTOR and
+      (attributes.direction == Direction.INPUT or attributes.direction == Direction.OUTPUT) and
+      not UnorderedSet.contains(variable.name, connectedIOs)
+    then
+      tlio_var := variable; // same attributes like start, unit
+      tlio_var.binding := UNBOUND(); // value is defined with tlio_eql
+      // find new name in global scope, using underscore instead of dot
+      cref := variable.name;
+      name := stringDelimitList(ComponentRef.toString_impl(cref, {}), "_");
+      while UnorderedMap.contains(tlio_var.name, variables) loop
+        tlio_node := InstNode.NAME_NODE(name);
+        tlio_var.name := match cref case ComponentRef.CREF() then
+          ComponentRef.CREF(tlio_node, cref.subscripts, cref.ty, cref.origin, ComponentRef.EMPTY());
```

Comment:
I'm not sure if it matters in this case, but this will only keep the subscripts from the last part of the cref. It might be a good idea to call `ComponentRef.combineSubscripts` on the cref before doing this.

Reply:
Good hint.